src/test/java/org/opensearch/security/transport/SecurityInterceptorTests.java

```diff
@@ -201,17 +292,306 @@ public <T extends TransportResponse> void sendRequest(
+     */
     @Test
-    public void testSendRequestDecorate() {
-        testSendRequestDecorate(Version.CURRENT);
+    public void testSendRequestDecorateWithNullFields() {
+        Version remoteNodeVersion = Version.CURRENT;
+        threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_USER, user);
+        boolean useJDKSerialization = remoteNodeVersion.before(ConfigConstants.FIRST_CUSTOM_SERIALIZATION_SUPPORTED_OS_VERSION);
+        ClusterName clusterName = ClusterName.DEFAULT;
+        when(clusterService.getClusterName()).thenReturn(clusterName);
+
+        MockTransport transport = new MockTransport();
+        TransportService transportService = transport.createTransportService(
+            Settings.EMPTY,
+            threadPool,
+            TransportService.NOOP_TRANSPORT_INTERCEPTOR,
+            boundTransportAddress -> clusterService.state().nodes().get(SecurityInterceptor.class.getSimpleName()),
+            null,
+            emptySet(),
+            NoopTracer.INSTANCE
+        );
+
+        // CS-SUPPRESS-SINGLE: RegexpSingleline Extensions manager used for creating a mock
+        OpenSearchSecurityPlugin.GuiceHolder guiceHolder = new OpenSearchSecurityPlugin.GuiceHolder(
+            mock(RepositoriesService.class),
+            transportService,
+            mock(IndicesService.class),
+            mock(PitService.class),
+            mock(ExtensionsManager.class)
+        );
+        // CS-ENFORCE-SINGLE
+
+        DiscoveryNode localNode = new DiscoveryNode("local-node1", new TransportAddress(localAddress, 1234), Version.CURRENT);
+        Connection connection1 = transportService.getConnection(localNode);
+
+        DiscoveryNode remoteNode = new DiscoveryNode("remote-node", new TransportAddress(localAddress, 6789), remoteNodeVersion);
+        Connection connection2 = transportService.getConnection(remoteNode);
+
+        // checking thread context inside sendRequestDecorate for connection1 with null local node -- we should serialize this
+        AsyncSender sender = new AsyncSender() {
+            @Override
+            public <T extends TransportResponse> void sendRequest(
+                Connection connection,
+                String action,
+                TransportRequest request,
+                TransportRequestOptions options,
+                TransportResponseHandler<T> handler
+            ) {
+                String serializedUserHeader = threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER);
+                assertEquals(serializedUserHeader, Base64Helper.serializeObject(user, useJDKSerialization));
+            }
+        };
+
+        // this is a request where the local node is null; have to use the remote connection since the serialization will fail
+        securityInterceptor.sendRequestDecorate(sender, connection2, action, request, options, handler, null);
```

Comment:
I'm not sure we know sender has executed, lets make sure this is confirmed by tying a completable future or countdownlatch to wait on before proceeding.

Reply:
Note: this is the only comment that I'd feel is a barrier to merging this change

Reply:
Sure will do. I think that the sender validation should theoretically check it but I will add this in case I am wrong.